subprojects/core/src/main/java/org/gradle/caching/internal/controller/impl/RootBuildCacheControllerRef.java

```diff
@@ -17,63 +17,179 @@
+            try {
+                controller.close();
+            } catch (IOException e) {
+                throw UncheckedException.throwAsUncheckedException(e);
+            }
+        }
+
+        protected abstract BuildCacheController getDelegate();
     }
 
-    private static class CloseShieldBuildCacheController implements BuildCacheController {
-        private final BuildCacheController delegate;
+    private static class RootBuildCacheController extends DelegatingBuildCacheController {
+        private final AtomicReference<BuildCacheController> delegate = new AtomicReference<>();
+        private Path identityPath;
+        private BuildCacheControllerFactory buildCacheControllerFactory;
+        private InstanceGenerator instanceGenerator;
 
-        private CloseShieldBuildCacheController(BuildCacheController delegate) {
-            this.delegate = delegate;
+        private RootBuildCacheController() {
+            this.delegate.set(NoOpBuildCacheController.INSTANCE);
         }
 
-        @Override
-        public boolean isEnabled() {
-            return delegate.isEnabled();
+        public RootBuildCacheController init(Path identityPath, BuildCacheControllerFactory buildCacheControllerFactory, InstanceGenerator instanceGenerator) {
+            this.identityPath = identityPath;
+            this.buildCacheControllerFactory = buildCacheControllerFactory;
+            this.instanceGenerator = instanceGenerator;
+            return this;
         }
 
         @Override
-        public Optional<BuildCacheLoadResult> load(BuildCacheKey cacheKey, CacheableEntity cacheableEntity) {
-            return delegate.load(cacheKey, cacheableEntity);
+        public void configurationAvailable(BuildCacheConfigurationInternal configuration) {
+            createDelegate(configuration, this.delegate, buildCacheControllerFactory, identityPath, instanceGenerator);
         }
 
         @Override
-        public void store(BuildCacheKey cacheKey, CacheableEntity entity, Map<String, FileSystemSnapshot> snapshots, Duration executionTime) {
-            delegate.store(cacheKey, entity, snapshots, executionTime);
+        protected BuildCacheController getDelegate() {
+            return delegate.get();
         }
 
         @Override
-        public void close() {
+        public void resetState() {
+            discardDelegate(delegate);
+        }
+
+        public LifecycleAwareBuildCacheController createChild(Path identityPath, BuildCacheControllerFactory buildCacheControllerFactory, InstanceGenerator instanceGenerator) {
+            return new NonRootBuildCacheController(this, identityPath, buildCacheControllerFactory, instanceGenerator);
         }
     }
 
+    private static class NonRootBuildCacheController extends DelegatingBuildCacheController {
+        private final RootBuildCacheController rootController;
+        private final AtomicReference<BuildCacheController> delegate = new AtomicReference<>();
```

Comment:
Holds a BCC that represents the current state of this build's cache configuration, but only if used as an 'early' build.

Reply:
These comments seem to be worth adding in code, too.

Reply:
Added